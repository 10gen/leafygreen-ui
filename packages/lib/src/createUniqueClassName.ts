// https://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid
// This is a RFC4122 version 4 compliant solution that solves that issue by offsetting the first 13 hex numbers by a hex portion of the timestamp, and once depleted offsets by a hex portion of the microseconds since pageload.
// That way, even if Math.random is on the same seed, both clients would have to generate the UUID the exact same number of microseconds since pageload (if high-perfomance time is supported) AND at the exact same millisecond (or 10,000+ years later) to get the same UUID:

const generateUUID = () => {
  let d = new Date().getTime(),
    d2 = (performance && performance.now && performance.now() * 1000) || 0;
  return 'xxxxxxxx'.replace(/[xy]/g, c => {
    let r = Math.random() * 16;
    if (d > 0) {
      r = (d + r) % 16 | 0;
      d = Math.floor(d / 16);
    } else {
      r = (d2 + r) % 16 | 0;
      d2 = Math.floor(d2 / 16);
    }
    return (c == 'x' ? r : (r & 0x7) | 0x8).toString(16);
  });
};

const createUniqueClassName = () => {
  return `lg-ui-${generateUUID()}`;
};

export default createUniqueClassName;
