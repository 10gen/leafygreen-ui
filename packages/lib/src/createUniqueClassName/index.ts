// https://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid
// This function returns a randomized 8-character string based on new Date()

const generateUUID = () => {
  let date1 = new Date().getTime();
  return 'xxxxxxxx'.replace(/[x]/g, char => {
    let rand = Math.random() * 16;
    rand = (date1 + rand) % 16 | 0;
    date1 = Math.floor(date1 / 16);
    return (char == 'x' ? rand : (rand & 0x7) | 0x8).toString(16);
  });
};

const createUniqueClassName = () => {
  return `lg-ui-${generateUUID()}`;
};

export default createUniqueClassName;
