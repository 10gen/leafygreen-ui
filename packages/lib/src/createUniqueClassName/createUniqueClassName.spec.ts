import createUniqueClassName from '.';

describe('packages/lib/createUniqueClassName', () => {
  describe('createUniqueClassName', () => {
    test('Format is correct', () => {
      const prefix = 'lg-ui-';
      const expectedUuidLength = 8;
      const result = createUniqueClassName();
      expect(result.startsWith(prefix)).toBe(true);
      expect(result.length == prefix.length + expectedUuidLength);
    });

    test('Two generated classNames are not equal.', () => {
      const res1 = createUniqueClassName();
      const res2 = createUniqueClassName();
      expect(res1.valueOf() == res2.valueOf()).not.toBe(true);
    });
  });
});
