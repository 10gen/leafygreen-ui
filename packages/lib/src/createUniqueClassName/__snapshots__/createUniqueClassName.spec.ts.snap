// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packages/lib/createUniqueClassName rendering Classnames persist in snapshot tests 1`] = `
<div>
  <div
    class="lg-ui-2a0d21fb"
    data-testid="el-1"
  >
    <div
      class="lg-ui-9222cd7a"
      data-testid="el-2"
    />
  </div>
</div>
`;
